File: dotnet/ArqLibrarianClassic/ArqLibrarianClassic/Application.cs
﻿using System.Collections.Generic;
using System.Linq;
using ArqLibrarianClassic.Library;

namespace ArqLibrarianClassic
{
    public class Application
    {
        private readonly UserIn input;
        private readonly UserOut output;
        private BooksManager booksManager;
        private BorrowingApplicationService borrowingManager;

        //for now application works for the only sample user with id = 1
        //it will be more dynamic when login process will be supported
        long userId = 1L;

        public Application(UserIn input, UserOut output)
        {
            this.input = input;
            this.output = output;
        }


        public bool Start()
        {
            output.PrintLine("Welcome to the ArqLibrarian");

            var running = true;

            while (running) {
                output.Print("> ");
                var commandString = input.ReadLine();
                var args = commandString.Split(' ');
                var commandName = args[0];

                switch (commandName)
                {
                    case "exit":
                        running = false;
                        break;
                    case "search":
                        SearchBook(commandString);
                        break;
                    case "rate":
                        RateBook(args);
                        break;
                    case "add":
                        AddBook();
                        break;
                    case "borrow":
                        BorrowBook(args);
                        break;
                    case "status":
                        ShowStatus(args);
                        break;
                    default:
                        output.PrintLine($"Comand {commandName} not recognized");
                        break;
                }
            }

            return running;
        }

        private void ShowStatus(string[] args)
        {
            var bookId = long.Parse(args[1]);

            bool borrowed = borrowingManager.Borrowed(bookId);
            var book = booksManager.FindById(bookId);
            
            output.PrintLine($"{BorrowedToString(borrowed)} {BasicInfoFor(book)}");
        }

        private void BorrowBook(string[] args)
        {
            var bookId = long.Parse(args[1]);
            borrowingManager.Borrow(userId, bookId);
            var book = booksManager.FindById(bookId);
            
            output.PrintLine($"Borrowed: {BasicInfoFor(book)}");
        }

        private void RateBook(string[] args)
        {
            if (args.Length != 3)
            {
                output.PrintLine("Wrong rate command format. Try: rate [book id] [rating]");
            }
            var bookId = long.Parse(args[1]);
            var rating = int.Parse(args[2]);

            booksManager.Rate(bookId, rating);
            var book = booksManager.FindById(bookId);
            var totalRating = booksManager.ComputeRatingFor(bookId);
            
            output.PrintLine($"{book.Title} rated: {rating}, total rating: {totalRating}");
        }

        private void SearchBook(string command)
        {  
            IEnumerable<Book> books = null;
            if (HasParameters(command))
            {
                books = booksManager.FindAll();
            }
            else
            {
                var toSearch = SubstringAfterFirstSpace(command);
  
                books = booksManager.FindByTitle(toSearch);
                output.PrintLine(books.Any() ? $"Found: '{toSearch}'" : $"'{toSearch}' title not Found");
            }

            Print(books);
        }

        private static bool HasParameters(string command)
        {
            return command.IndexOf(' ') < 0;
        }

        private static string SubstringAfterFirstSpace(string command)
        {
            return command.Substring(command.IndexOf(' ') + 1);
        }

        private void Print(IEnumerable<Book> books)
        {
            foreach (var book in books)
            {
                var rating = booksManager.ComputeRatingFor(book.Id);
                output.PrintLine($"{book.Id}: {BasicInfoFor(book)}, rating: {rating}");
            }
        }

        private string BorrowedToString(bool borrowed)
        {
            return borrowed ? "[borrowed]" : "[available]";
        }

        private string BasicInfoFor(Book book)
        {
            return $"\"{book.Title}\" - {book.Author}, {book.Category}";
        }

        private void AddBook()
        {
            output.PrintLine("Adding a new book");
            output.Print("Title: ");
            var title = input.ReadLine();
            
            output.Print("Author: ");
            var author = input.ReadLine();

            output.Print("isbn: ");
            var isbn = input.ReadLine();
            
            output.Print("Publihser: ");
            var publisher = input.ReadLine();
            
            output.Print("Year: ");
            var yearString = input.ReadLine();
            var year = int.Parse(yearString);
            
            output.Print("Category: ");
            var category = input.ReadLine();

            booksManager.Create(title, author, isbn, publisher, year, category);
        }

        public void Setup(BooksManager manager)
        {
            this.booksManager = manager;
        }

        public void Setup(BorrowingApplicationService manager)
        {
            this.borrowingManager = manager;
        }
    }
}

File: java/ArqLibraryClassic/ArqLibraryLayered/src/bnsit/ms/arq/Application.java
package bnsit.ms.arq;

import bnsit.ms.arq.library.Book;
import bnsit.ms.arq.library.BooksManager;
import bnsit.ms.arq.library.BorrowingManager;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Application {

    private UserIn input;
    private UserOut output;
    private BooksManager booksManager;
    private BorrowingManager borrowingManager;

    //for now application works for the only sample user with id = 1
    //it will be more dynamic when login process will be supported
    private long userId = 1L;

    public Application(UserIn input, UserOut output) {

        this.input = input;
        this.output = output;
    }

    public void setup(BooksManager booksManager) {
        this.booksManager = booksManager;
    }

    public void setup(BorrowingManager borrowingManager) {
        this.borrowingManager = borrowingManager;
    }

    public void start() {
        output.printLine("Welcome to the ArqLibrarian");

        boolean running = true;

        while (running) {
            output.print("> ");
            String commandString = input.readLine();
            String[] args = commandString.split(" ");
            String commandName = args[0];

            switch (commandName)
            {
                case "exit":
                    running = false;
                    break;
                case "search":
                    this.searchBook(commandString);
                    break;
                case "rate":
                    this.rateBook(args);
                    break;
                case "add":
                    this.addBook();
                    break;
                case "borrow":
                    this.borrowBook(args);
                    break;
                case "status":
                    this.showStatus(args);
                    break;
                default:
                    output.printLine(String.format("Command %s not recognized", commandName));
                    break;
            }
        }
    }

    private void showStatus(String[] args) {
        long bookId = Long.parseLong(args[1]);

        boolean borrowed = borrowingManager.borrowed(bookId);
        Book book = booksManager.findById(bookId);

        output.printLine(String.format("%s %s", borrowedToString(borrowed), basicInfoFor(book)));
    }

    private String borrowedToString(boolean borrowed) {
        return borrowed ? "[borrowed]" : "[available]";
    }

    private void borrowBook(String[] args) {
        long bookId = Long.parseLong(args[1]);
        borrowingManager.borrow(userId, bookId);
        Book book = booksManager.findById(bookId);

        output.printLine(String.format("Borrowed: %s", basicInfoFor(book)));
    }

    private void addBook() {
        output.printLine("Adding a new book");
        output.print("Title: ");
        String title = input.readLine();

        output.print("Author: ");
        String author = input.readLine();

        output.print("isbn: ");
        String isbn = input.readLine();

        output.print("Publihser: ");
        String publisher = input.readLine();

        output.print("Year: ");
        String yearString = input.readLine();
        int year = Integer.parseInt(yearString);

        output.print("Category: ");
        String category = input.readLine();

        booksManager.create(title, author, isbn, publisher, year, category);
    }

    private void rateBook(String[] args) {
        if (args.length != 3)
        {
            output.printLine("Wrong rate command format. Try: rate [book id] [rating]");
        }
        long bookId = Long.parseLong(args[1]);
        int rating = Integer.parseInt(args[2]);

        booksManager.rate(bookId, rating);
        Book book = booksManager.findById(bookId);
        double totalRating = booksManager.computeRatingFor(bookId);

        output.printLine(String.format("%s rated: %d, total rating: %1.1f", book.getTitle(), rating, totalRating));
    }

    private void searchBook(String commandString) {
        Iterator<Book> books = null;
        if (hasParameters(commandString))
        {
            books = booksManager.findAll();
        }
        else
        {
            String toSearch = substringAfterFirstSpace(commandString);

            books = booksManager.findByTitle(toSearch);
            output.printLine(books.hasNext()
                    ? String.format("Found: '%s'", toSearch)
                    : String.format("'%s' title not Found", toSearch));
        }

        this.print(books);
    }

    private void print(Iterator<Book> iterator) {
        List<Book> books = new ArrayList<>();
        iterator.forEachRemaining(books::add);

        for(Book book : books)
        {
            double rating = booksManager.computeRatingFor(book.getId());
            output.printLine(String.format("%s: %s, rating: %s", book.getId(), basicInfoFor(book), formatted(rating)));
        }
    }

    private String formatted(double rating) {
        if (rating < 0.0) {
            return "Not rated yet";
        }

        return String.format("%1.1f", rating);
    }

    private String basicInfoFor(Book book) {
        return String.format("\"%s\" - %s, %s", book.getTitle(), book.getAuthor(), book.getCategory());
    }

    private static String substringAfterFirstSpace(String command)
    {
        return command.substring(command.indexOf(' ') + 1);
    }

    private static boolean hasParameters(String command)
    {
        return command.indexOf(' ') < 0;
    }
}
